include/base_loop.h

```diff
@@ -0,0 +1,213 @@
+/**
+ * @brief RL Inference API definition.
+ *
+ * @file base_loop.h
+ * @author Rajan Chari et al
+ * @date 2018-07-18
+ */
+#pragma once
+#include "err_constants.h"
+#include "factory_resolver.h"
+#include "future_compat.h"
+#include "sender.h"
+
+#include <cstring>
+#include <functional>
+#include <memory>
+
+#define INIT_CHECK()                                                                                     \
+  do {                                                                                                   \
+    if (!_initialized)                                                                                   \
+    {                                                                                                    \
+      RETURN_ERROR_ARG(nullptr, status, not_initialized, "Library not initialized. Call init() first."); \
+    }                                                                                                    \
+  } while (0);
+
+namespace reinforcement_learning
+{
+//// Forward declarations ////////
+class live_model_impl;   //
+class ranking_response;  //
+class api_status;        //
+                         //
+namespace model_management
+{                        //
+class i_data_transport;  //
+class i_model;           //
+}  // namespace model_management
+   //
+namespace utility
+{                     //
+class configuration;  //
+}  // namespace utility
+//////////////////////////////////
+
+struct loop_str
```

Comment:
I don't understand why our C++ interface must become a C interface? 

C++ has worked for many years. If we require a pure C interface that we want to provide for ABI stability and binding convenience then it is trivial to layer it upon this. But by dropping our core interface to C we lose so much expressiveness and productivity.

Reply:
structs are C++...

Keeping the interface surface area minimal using appropriate types for its intended purpose makes  it a good C++ interface.

Here the intent is to pass in buffer and size as one logical entity.  I think this choice serves that purpose.